File: solutions/rust/0056.rs
impl Solution {
    pub fn merge(intervals: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        /*
         * 合并区间
         *
         * 时间复杂度: O(n log n)，其中 n 为区间数量（排序耗时）
         * 空间复杂度: O(n)，用于结果存储（原地排序）
         *
         * 思路:
         * 1. 按区间起点排序
         * 2. 线性扫描，用 current 表示当前合并区间
         *    - 若下一个区间的起点 > current 终点，则无重叠：推入结果，重置 current
         *    - 否则有重叠：更新 current 终点 = max(current 终点, 下个终点)
         * 3. 扫描结束推入最后的 current
         */
        let mut intervals = intervals;
        if intervals.is_empty() {
            return Vec::new();
        }

        // 按起点排序
        intervals.sort_by_key(|itv| itv[0]);

        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut current = intervals[0].clone(); // [start, end]

        for itv in intervals.into_iter().skip(1) {
            let start = itv[0];
            let end = itv[1];
            let cur_end = current[1];

            if start > cur_end {
                // 不重叠，先保存 current，再开始新的区间
                result.push(current);
                current = vec![start, end];
            } else {
                // 重叠，扩展 current 的终点
                current[1] = cur_end.max(end);
            }
        }

        // 推入最后一个合并后的区间
        result.push(current);

        result
    }
}

File: solutions/go/0056.go
import "sort"

func merge(intervals [][]int) [][]int {
	/*
	 * 合并区间
	 *
	 * 时间复杂度: O(n log n)，其中 n 为区间数量
	 *   - 主要耗时在对区间按起点排序
	 * 空间复杂度: O(n)，用于存储结果数组
	 *
	 * 思路:
	 * 1. 先按每个区间的起点从小到大排序
	 * 2. 遍历排序后的区间数组，用一个结果数组保存合并后的区间
	 * 3. 对于当前区间 curr:
	 *    - 如果结果数组为空，或 curr 的起点 > 结果数组最后一个区间的终点，说明没有重叠，直接加入结果
	 *    - 否则存在重叠，更新最后一个区间的终点为 max(原终点, curr 的终点)
	 * 4. 遍历结束后，结果数组即为合并后的不重叠区间集合
	 *
	 * 示例:
	 * intervals = [[1,3],[2,6],[8,10],[15,18]]
	 * 排序后: [[1,3],[2,6],[8,10],[15,18]]
	 * 合并过程:
	 *   - 结果: [[1,3]]
	 *   - curr=[2,6] 与 [1,3] 重叠，合并为 [1,6]
	 *   - curr=[8,10] 与 [1,6] 不重叠，加入 => [[1,6],[8,10]]
	 *   - curr=[15,18] 与 [8,10] 不重叠，加入 => [[1,6],[8,10],[15,18]]
	 */
	if len(intervals) == 0 {
		return [][]int{}
	}

	// 按起点从小到大排序
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})

	result := make([][]int, 0)
	// 初始化为第一个区间（创建副本）
	current := []int{intervals[0][0], intervals[0][1]}

	for i := 1; i < len(intervals); i++ {
		start := intervals[i][0]
		end := intervals[i][1]
		curEnd := current[1]

		// 没有重叠，推入当前区间，并开始新的区间
		if start > curEnd {
			result = append(result, current)
			current = []int{start, end}
		} else {
			// 有重叠，更新当前区间的结束位置
			if end > curEnd {
				current[1] = end
			}
		}
	}

	// 别忘了最后一个区间
	result = append(result, current)

	return result
}